File: NN/js/que.js
// @ts-check
"use strict";

class Que {
    constructor(brain) {

        /*let input1 = random();
        let input2 = random();

        this.angle = map(input1, 0, 1, 0, PI);
        this.force = map(input2, 0, 1, 0.5, 1);*/
        this.angle = 0;
        this.force = 0;
        this.initialForce = 0;

        this.resetResults();

        /*
        if (brain) {
            this.brain = brain.copy();
        } else {
            this.brain = new NeuralNetwork(6, 8, 2);
        }*/
    }

    postAjax(url, data, success) {

        let xhr = new XMLHttpRequest();
        xhr.open("POST", url, true);
        xhr.setRequestHeader("Content-type", "application/json");
        xhr.onreadystatechange = function () { 
            if (xhr.readyState == 4 && xhr.status == 200) {
                var json = JSON.parse(xhr.responseText);
                this.angle = json.angle;
                this.force = json.force
                this.initialForce = this.force;
                success();
            }
        };

        var jsonstring = JSON.stringify(data);
        xhr.send(jsonstring);
        return xhr;
    }

    thinkPython(whiteBall, redBall, hole, doneCallback) {
        let data = {}
        data.whiteBall ={ x: whiteBall.pos.x, y: whiteBall.pos.y }
        data.redBall ={ x: redBall.pos.x, y: redBall.pos.y }
        data.hole ={ x: hole.pos.x, y: hole.pos.y }
        this.postAjax("/predict", data, doneCallback)
    }

    think(whiteBall, redBall, hole) {
        let inputs = [];
        inputs[0] = whiteBall.pos.x / width;
        inputs[1] = whiteBall.pos.y / height;
        inputs[2] = redBall.pos.x / width;
        inputs[3] = redBall.pos.y / height;
        inputs[4] = hole.pos.x / width;
        inputs[5] = hole.pos.y / height;
        let output = this.brain.predict(inputs);
        // console.log(output);
        this.angle = map(output[0], 0, 1, 0, PI);
        this.force = map(output[1], 0, 1, 0.5, 1);
        this.initialForce = this.force;
    }

    resetResults()
    {
        this.ballsHit = false;
        this.redPotted = false;
        this.whitePotted = false;
        this.score = 0;
        this.closestDistance = width;
    }

    mutate() {
        this.brain.mutate(0.1);
    }

    moveTo(x, y) {
        this.x = x;
        this.y = y;
    }

    getForce() {
        let w = p5.Vector.fromAngle(this.angle - PI / 2);
        w.setMag(this.initialForce * 10);
        return w;
    }

    updateScore(hole, ball) {
        var dist = ball.pos.dist(hole.pos);
        if (dist < this.closestDistance) {
            this.closestDistance = dist;
        }
    }

    calcScore() {
        if (this.redPotted) {
            this.score = 1
        }
        else if (this.whitePotted) {
            this.score = 0;
        }
        else if (this.ballsHit === false) {
            this.score = 0;
        }
        else {
            this.score = 1 - (this.closestDistance / width);
        }

        return this.score;
    }

    draw() {
        strokeWeight(8);
        stroke(150, 75, 0);

        let v = createVector(this.x, this.y);
        let w = p5.Vector.fromAngle(this.angle - PI / 2);
        w.setMag(15 + (this.force > 0 ? 20 : 0));
        let p1 = p5.Vector.sub(v, w);
        w.setMag(100);
        let p2 = p5.Vector.sub(v, w);

        line(p1.x, p1.y, p2.x, p2.y);

        strokeWeight(1);
    }
}

if (typeof module !== 'undefined') {
    module.exports = Que;
}

File: NN/Que.py
import numpy as np
from Vector2 import Vector2
from NeuralNetwork import NeuralNetwork
from Ball import Ball
from Hole import Hole
import Utils
import Config
import math

class Que():

	x = 0
	y = 0
	angle = 0
	force = 0
	initialForce = 0

	def __init__(self, brain = None):
		self.angle = 0
		self.force = 0
		self.initialForce = 0

		self.ballsHit = False
		self.redPotted = False
		self.whitePotted = False
		self.totalScore = 0
		self.closestDistance = Config.WIDTH
		self.gamesPlayed = 0

		if brain != None:
				self.brain = brain.copy()
		else: 
				# self.brain = NeuralNetwork(6, Config.HIDDEN_LAYERS, Config.HIDDEN_NODES, 2)
				self.brain = NeuralNetwork(3, Config.HIDDEN_LAYERS, Config.HIDDEN_NODES, 2)

	def copy(self):
		c = Que(self.brain)
		c.ballsHit = self.ballsHit
		c.redPotted = self.redPotted
		c.whitePotted = self.whitePotted
		c.totalScore = self.totalScore
		c.gamesPlayed = self.gamesPlayed
		c.closestDistance = self.closestDistance
		return c

	def think(self, whiteBall, redBall, hole):
		wx = whiteBall.pos.x / Config.WIDTH
		wy = whiteBall.pos.y / Config.HEIGHT
		rx = redBall.pos.x / Config.WIDTH
		ry = redBall.pos.y / Config.HEIGHT
		hx = hole.pos.x / Config.WIDTH
		hy = hole.pos.y / Config.HEIGHT
		# inputs = np.array([wx, wy, rx, ry, hx, hy])
		inputs = np.array([wy, ry, hy])
		output = self.brain.predict(inputs)
		# print(output)
		self.angle = Utils.map(output[0], 0, 1, 0, math.pi, False)
		self.force = Utils.map(output[1], 0, 1, 0.5, 1, False)
		self.initialForce = self.force

	def mutate(self):
		self.brain.mutate(0.1)

	def move_to(self, x, y):
		self.x = x
		self.y = y

	def get_force(self):
		w = Vector2.fromAngle(self.angle - math.pi / 2, 1)
		w.setMag(self.initialForce * 10)
		return w

	def update_score(self, hole, ball):
		dist = ball.pos.dist(hole.pos)
		if dist < self.closestDistance:
			self.closestDistance = dist

	def apply_score(self):
		score = 0
		if self.redPotted:
			score = 1
		elif self.whitePotted:
			score = 0
		elif self.ballsHit == False:
			score = 0
		else:
			score = (1 - (self.closestDistance / Config.WIDTH))

		self.totalScore += score;
		self.gamesPlayed += 1

		self.redPotted = False
		self.whitePotted = False
		self.ballsHit = False
		self.closestDistance = Config.WIDTH

		return self.totalScore / self.gamesPlayed

